src/main/java/com/exasol/glue/ExasolTable.java

```diff
@@ -47,36 +42,74 @@ public class ExasolTable implements SupportsRead {
      */
     public ExasolTable(final StructType schema) {
         this.schema = schema;
-        this.capabilities = new HashSet<>(Arrays.asList(TableCapability.BATCH_READ));
+        this.capabilities = new HashSet<>(Arrays.asList(TableCapability.BATCH_READ, TableCapability.BATCH_WRITE));
     }
 
     @Override
+    // [impl->dsn~sourcescanbuilder-prunes-columns-and-pushes-filters~1]
     public ScanBuilder newScanBuilder(final CaseInsensitiveStringMap map) {
         final ExasolOptions options = getExasolOptions(map);
-        final S3ClientFactory s3ClientFactory = new S3ClientFactory(options);
-        final SparkSession sparkSession = SparkSession.active();
-        final String s3Bucket = options.getS3Bucket();
-        final String s3BucketKey = UUID.randomUUID() + "-" + sparkSession.sparkContext().applicationId();
-        LOGGER.info(() -> "Using bucket '" + s3Bucket + "' with folder '" + s3BucketKey + "' for job data.");
-        validateS3BucketExists(s3ClientFactory, s3Bucket);
-        runExportQuery(options, s3BucketKey);
-        setupSparkContextForS3(sparkSession, options);
-        setupSparkContextForJobListener(sparkSession, options, s3BucketKey);
-        final List<String> path = Arrays.asList(getS3Path(s3Bucket, s3BucketKey));
-        final CSVTable csvTable = new CSVTable("", //
-                sparkSession, //
-                map, //
-                JavaConverters.asScalaIteratorConverter(path.iterator()).asScala().toSeq(), //
-                Option.apply(this.schema), //
-                null);
-        return csvTable.newScanBuilder(updateMapWithCSVOptions(map));
+        validate(options);
+        updateSparkConfigurationForS3(options);
```

Comment:
Is this modifying global state? If so, that would be dangerous in a concurrent environment.

Reply:
Yes, good point. Add synchronized block on configuration object.